initrd/etc/functions

```diff
@@ -132,66 +132,85 @@ enable_usb_storage() {
+		while read blkdev_dir
+		do # Iterate over each detected block device to apply filters:
+			blkdev="/dev/$(basename "$blkdev_dir")" # Get block device path from directory basename
+			if [ ! -b "$blkdev" ]
+			then # Ignore this block device directory: cannot identify associated block device
+				DEBUG "Ignoring \"${blkdev_dir}\": could not identify associated block device"
+				continue
+			elif [ ! -f "${blkdev_dir}/size" ] || [ "$(cat "${blkdev_dir}/size")" -le '0' ]
+			then # Ignore this device with a size of '0' (may be an empty USB SD card reader, for example)
+				DEBUG "Ignoring USB block device \"${blkdev}\" with a size of 0"
+				continue
+			fi
+
+			if [ "${1-}" = 'devices' ]
+			then # Include block device, skipping over additional (partition) checks:
+				DEBUG "Including USB block device \"${blkdev}\" due to \"devices\" argument"
+				echo "$blkdev"
+				continue
 			fi
-		done |
-		sed "s|/sys/block|/dev|" |
-		while read b; do
-			# If the device has a partition table, ignore it and
-			# include the partitions instead - even if the kernel
-			# hasn't detected the partitions yet.  Such a device is
-			# never usable directly, and this allows the "wait for
-			# disks" loop in mount-usb to correctly wait for the
-			# partitions.
-			# This check: [ $(fdisk -l "$b" | wc -l) -eq 5 ]
-			# covers the case of a device without partition table but
-			# formatted as fat32, which contains a sortof partition table.
-			# this causes fdisk to not print the invalid partition table
-			# message and instead it'll print an empty table with header.
-			# In both cases the output is 5 lines: 3 about device info,
-			# 1 empty line and the 5th will be the table header or the
-			# unvalid message.
-			DISK_DATA=$(fdisk -l "$b")
-			if echo "$DISK_DATA" | grep -q "doesn't contain a valid partition table" || [ $(echo "$DISK_DATA" | wc -l) -eq 5 ]; then
-				# No partition table, include this device
-				DEBUG "USB storage device without partition table: $b"
-				echo "$b"
-			#Bypass the check for partitions if we want only disks
-			elif [ "$1" = "disks" ]; then
-				# disks only were requested, so we don't list partitions
-				DEBUG "USB storage device with partition table: $b"
-				DEBUG "We asked for disks only, so we don't want to list partitions"
-				echo "$b"
-			else
-				# Has a partition table, include partitions
-				DEBUG "USB storage device with partition table: $b"
-				ls -1 "$b"* | awk 'NR!=1 {print $0}'
+
+			# Check if the USB block device has had a filesystem written directly
+			# to it, and include the device in the outputted list if so. Although
+			# this will not apply to the majority of block devices, it covers
+			# important edge cases, such as when bootable media has had an ISO
+			# created using the "isohybrid" utility, written to it.
+			test_mntpt='/tmp/direct-mount-test'
+			mkdir -p "$test_mntpt"
+			blkdev_included='false'
+			if mount -r "$blkdev" "$test_mntpt" >/dev/null 2>/dev/null
```

Comment:
Per my other comment, we need to go back to the strategy of checking if the device has a partition table first, and using only the partitions if it does.

I liked the idea of doing a mount test - maybe we could apply that to the partitions or to unpartitioned media in the future - but since it can apparently find remnant filesystems after repartitioning I don't think we can go this route.